chain-impl-mockchain/src/rewards.rs

```diff
@@ -163,6 +174,24 @@ pub fn rewards_contribution_calculation(epoch: Epoch, params: &Parameters) -> Va
     }
 }
 
+/// Added new function to ensure rewards calculation complies with research -- KH
+/// Scale the total rewards in proportion to the total delegated stake
+pub fn rewards_contribution_calculation_scaled(
+    epoch: Epoch,
+    params: &Parameters,
+    total_delegated: Stake,
+) -> Value {
+    let r = rewards_contribution_calculation(epoch, params);
+
+    // limit the rewards if necessary
+    let rl = &params.rewards_limit;
+    const SCALE: u128 = 10 ^ 18;
```

Comment:
That's a `xor` not a `power` function.

Reply:
OK.  This seems to be consistent with the previous uses.  Are there some bugs there?  I will fix them if so.  I see you had spotted the epoch=0 error for delegators that was previously there.

Reply:
This was copied from the existing Rust code. I see you've now fixed the bugs that you had.  I have made this consistent.